pydatastructs/graphs/algorithms.py

```diff
@@ -548,3 +549,45 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(graph, start, path_length, operation):
+    '''
+    Executes random walk from the starting node and given graph.
+
+    Parameters
+    ==========
+
+    graph: Graph
+        The graph on which random walk needs to be executed.
+    start: GraphNode or it's subclasses
+        The node/vertex from where random walk will begin.
+    path_length:
+        The maximum number of nodes/vertices traversed by random walk.
+
+    Returns
+    =======
+
+    path: list
+        Python list of nodes after execution of random walk.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import Graph, AdjacencyListGraphNode
+    >>> from pydatastructs import random_walk
+    >>> v1, v2, v3 = [AdjacencyListGraphNode(i) for i in range(3)]
+    >>> g = Graph(v1, v2, v3)
+    >>> g.add_edge(v1.name, v2.name)
+    >>> g.add_edge(v2.name, v3.name)
+    >>> g.add_edge(v3.name, v1.name)
+
+    >>> random_walk = random_walk(g, v1, 5, )
+    >>> random_walk == ['0', '1', '2', '0', '1']
+    True
+
+    '''
+
+    path = operation(graph, start, path_length)
```

Comment:
This was not what I was talking about. Please see the implementation of `breadth_first_search`. `operation` should take, `curr_node`, `prev_node` and more stuff as it has been done in BFS.

Reply:
In BFS, there is parent-child relation between current node and the next node, where `parent[next_node] == curr_node`. In `random walk`, if we have `operation`, we don't need the `next_node`( as we choose it randomly from neighbours), and the previous node can be found by `path[-2]`,(` path[-1]` being `curr_node`). I am not getting the analogy.

Reply:
Well, AFAICT, current node is just where you are. Tracing previous node is simple. Now, if you previous and current node are same then you found a loop. Moreover, we will not form a path, that can be done through the operation we are passing. operation will append nodes to the current path. That's just two examples of how we can use operation in random walks. So just the method of choosing next node is different. I hope you are getting my point.

Reply:
I pushed a commit which implements `loop_erased_random_walk`.
Also I have verified that `random.shuffle()` follows approximately uniform distribution as it follows
`fischer yates shuffle`.